osquery/main/shell.cpp

```diff
@@ -83,6 +89,44 @@ int profile(int argc, char *argv[]) {
   return 0;
 }
 
+// readline completion expects strings to be malloced. readline will free them
+// later.
+char *copy_string(const std::string &str) {
+  char *copy = (char *)malloc(str.size() + 1);
+  if (copy == nullptr) {
+    fprintf(stderr,
+            "Memory allocation failed during shell autocompletion. Exiting!");
+    osquery::Initializer::shutdown(EXIT_FAILURE);
+  }
+  strncpy(copy, str.c_str(), str.size() + 1);
+  return copy;
+}
+
+char *completion_generator(const char *text, int state) {
+  static std::vector<std::string> tables;
+  static size_t index;
+
+  if (state == 0) {
+    // new completion attempt
+    tables = osquery::Registry::names("table");
+    index = 0;
+  }
+
+  while (index < tables.size()) {
+    std::string table = tables[index];
```

Comment:
Yes, this can and should be a const reference.

I don't think there should be a need for a guard around this section. The [readline docs](http://web.mit.edu/gnu/doc/html/rlman_2.html#SEC36) don't mention anything, and if the callbacks are running in a single thread I don't anticipate any problems.